spec/new_integration/router/using_middleware_spec.rb

```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
+
+require "rack/test"
+
+RSpec.describe "Router with middleware", :application_integration do
+  include Rack::Test::Methods
+
+  let(:app) { Hanami.app }
+
+  specify "Middleware blah" do
+    with_tmp_directory(Dir.mktmpdir) do
+      write "config/application.rb", <<~RUBY
+        require "hanami"
+
+        module TestApp
+          class Application < Hanami::Application
+            config.logger = {stream: File.new("/dev/null", "w")}
+          end
+        end
+      RUBY
+
+      write "config/routes.rb", <<~RUBY
+        module TestApp
+          class Middleware
+            def self.called_as
+              @called_as ||= Hash.new { |hsh, key| hsh[key] = [] }
+            end
+
+            def initialize(app, tag)
+              @app = app
+              @tag = tag
+            end
+
+            def call(env)
+              self.class.called_as[@tag] << env["PATH_INFO"]
+              @app.call(env)
+            end
+          end
+        end
+
+        Hanami.application.routes do
+          slice :main, at: "/" do
+            # With this here, it ends up applying to _all_ routes, including admin
+            use TestApp::Middleware, :main
+
+            root to: "home#show"
+          end
+
+          slice :admin, at: "/admin" do
+            # With this here, all admin routes return 404
+            use TestApp::Middleware, :admin
```

Comment:
And obviously this shouldn't make all the admin routes become 404s!

Reply:
I encountered this recently. admin routes become nested once more.
* `/admin/admin` => admin's home#show
* `/admin/admin/test` => admin's test#show

And main routes are nested in too, causing
* `/admin` => main's home#show

I found it's because the `Hanami::Application::Routing::Middleware::Stack` creates a map for '/admin' but run the whole app under it.